edu.wpi.first.wpilib.plugins.riolog/src/netconsole3/RioConnector.java

```diff
@@ -0,0 +1,256 @@
+
+    // wait for a connection attempt to be successful, or timeout
+    lock.lock();
+    try {
+      while (!done && !attempts.isEmpty()) {
+        if (!cvDone.await(TOTAL_TIMEOUT_SEC, TimeUnit.SECONDS)) {
+          //logger.log("Connection timed out.");
+          return null;
+        }
+      }
+
+      if (socket == null) {
+        //logger.log("Not connected to robot");
+        return null;
+      }
+
+      logger.log("Connected to robot at " + socket.getInetAddress().getHostAddress());
+      return socket;
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private void startDsConnect() {
+    startAttempt("DS", new Thread(() -> {
+      try {
+        // Try to connect to DS on the local machine
+        Socket socket = new Socket();
+        InetAddress dsAddress = InetAddress.getByAddress(new byte[] {127, 0, 0, 1});
+        socket.connect(new InetSocketAddress(dsAddress, 1742), CONNECTION_TIMEOUT_MS);
+
+        // Read JSON "{...}".  This is very limited, does not handle
+        // quoted "}" or nested {}, but is sufficient for this purpose.
+        InputStream ins = new BufferedInputStream(socket.getInputStream());
+        ByteArrayOutputStream buf = new ByteArrayOutputStream();
+
+        int b;
+        // Throw away characters until {
+        while ((b = ins.read()) >= 0 && b != '{' && !isDone()) {}
+
+        // Read characters until }
+        while ((b = ins.read()) >= 0 && b != '}' && !isDone()) {
+          buf.write(b);
+        }
+
+        if (isDone()) {
+          return;
+        }
+
+        String json = buf.toString("UTF-8");
+
+        // Look for "robotIP":12345, and get 12345 portion
+        Matcher m = dsPattern.matcher(json);
+        if (!m.find()) {
+          logger.log("DS did not provide robotIP");
+          return;
+        }
+        long ip = 0;
+        try {
+          ip = Long.parseLong(m.group(1));
+        } catch (NumberFormatException e) {
+          if (!isDone()) {
+            logger.log("DS provided invalid IP: \"" + m.group(1) + "\"", e);
+          }
+        }
+
+        // If zero, the DS isn't connected to the robot
+        if (ip == 0) {
+          return;
+        }
+
+        // Kick off connection to that address
+        InetAddress address = InetAddress.getByAddress(new byte[] {
+            (byte)((ip >> 24) & 0xff),
+            (byte)((ip >> 16) & 0xff),
+            (byte)((ip >> 8) & 0xff),
+            (byte)(ip & 0xff)});
+        logger.log("DS provided " + address.getHostAddress());
+        startConnect(address);
+      } catch (Exception e) {
+        if (!isDone()) {
+          //logger.log("could not get IP from DS");
+        }
+      } finally {
+        finishAttempt("DS");
+      }
+    }));
+  }
+
+  // Address resolution can take a while, so we do this in a separate
+  // thread, and then kick off a connect attempt on every IPv4 that
+  // resolves.
+  private void startConnect(String host) {
+    startAttempt(host, new Thread(() -> {
```

Comment:
Needs to be a daemon thread for same reason as above.

Reply:
It is. See startAttempt().